src/dataset-builder/DatasetBuilder.ts

```diff
@@ -665,23 +653,48 @@ export const DatasetBuilderContents = ({
       requestValid &&
         h(ActionBar, {
           prompt: h(Fragment, [
-            datasetRequestParticipantCount.status === 'Ready'
-              ? formatCount(datasetRequestParticipantCount.state.result.total)
+            snapshotRequestParticipantCount.status === 'Ready'
+              ? formatCount(snapshotRequestParticipantCount.state.result.total)
               : h(Spinner),
             ' participants in this dataset',
           ]),
-          actionText: 'Request access to this dataset',
-          onClick: () => setRequestingAccess(true),
+          actionText: 'Request this data snapshot',
+          onClick: () => {
+            setSnapshotAccessRequest(
+              withErrorReporting('Error creating dataset request')(
+                async () =>
+                  await DataRepo()
+                    .snapshotAccessRequest()
+                    .createSnapshotAccessRequest(
+                      createSnapshotAccessRequest(
+                        '',
+                        '',
+                        snapshotId,
+                        cohorts,
+                        conceptSets,
+                        _.map(
+                          (valuesSet: HeaderAndValues<DatasetBuilderValue>) => ({
+                            domain: valuesSet.header,
+                            values: valuesSet.values,
+                          }),
+                          selectedValues // convert from HeaderAndValues<DatasetBuilderType>[] to ValueSet[]
+                        )
+                      )
+                    )
+              )
+            );
+            setRequestingAccess(true);
+          },
         }),
     ]),
-    requestingAccess &&
-      h(RequestAccessModal, {
-        cohorts: allCohorts,
-        conceptSets: allConceptSets,
-        valueSets: selectedValues,
-        onDismiss: () => setRequestingAccess(false),
-        snapshotId,
-      }),
+    snapshotAccessRequest.status === 'Loading'
+      ? spinnerOverlay
+      : snapshotAccessRequest.status === 'Ready' &&
+        requestingAccess &&
+        h(RequestAccessModal, {
+          onDismiss: () => setRequestingAccess(false),
+          snapshotId: snapshotAccessRequest.state.id,
+        }),
   ]);
```

Comment:
what's the best way of doing this? Could I remove the 'requestingAccess', because if there is an issue an error will pop up, but if status === 'Ready' we can assume that `setRequestingAccess(true)` has been called

Reply:
This might be a place to do `Utils.switchCase(snapshotAccessRequest.status, ...` - that way we can avoid the nested ternaries.